File: app/src/main/java/su/linka/pictures/SetsManager.java
package su.linka.pictures;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import net.lingala.zip4j.ZipFile;
import net.lingala.zip4j.exception.ZipException;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;
import java.util.Comparator;
import java.util.UUID;

public class SetsManager {

    private final Cookie cookie;
    private final Context context;

    public SetsManager(Context context){
        this.context = context;
        cookie = new Cookie(context);
    }

    public void loadDefaultSets() throws IOException {
        if(cookie.get(Cookie.ASSETS_LOADER, false)){
            return;
        }
        File sets = getSetsDirectory();
        sets.mkdir();
        AssetManager am = getAssetsManager();
            String[] list = am.list("sets");
            for (int i = 0; i < list.length; i++) {
                InputStream in = am.open("sets/"+list[i]);
                File outFile = new File(sets.getAbsoluteFile(), list[i]);

                OutputStream out = new FileOutputStream(outFile);
                byte[] buffer = new byte[1024];
                int read;
                while ((read = in.read(buffer)) != -1) {
                    out.write(buffer, 0, read);
                }
                in.close();
                out.close();
            }
      cookie.set(Cookie.ASSETS_LOADER, true);
    }

    public File getSetsDirectory() {
        return new File(getRootDirectory(). getAbsoluteFile(), "/sets/");
    }


    private File getRootDirectory(){
        //Create File object for Parent Directory
        return context.getFilesDir();
    }

    private AssetManager getAssetsManager(){
        return context.getAssets();
    }

    public SetManifest[] getSets() {

        File[] files = getSetsDirectory().listFiles();
        Arrays.sort(files, new Comparator<File>() {
            @Override
            public int compare(File o1, File o2) {

                if (o1.lastModified() > o2.lastModified()) {
                    return -1;
                } else if (o1.lastModified() < o2.lastModified()) {
                    return +1;
                } else {
                    return 0;
                }
            }

        });
        SetManifest[] manifests = new SetManifest[files.length];
        for (int i = 0; i < files.length; i++) {
            File file = files[i];
            try {

               manifests[i] =  getSetManifest(file);
            } catch (IOException e) {
                Log.e(getClass().getCanonicalName(), "getSets: ", e);
                e.printStackTrace();
            }
        }
        return manifests;
    }
    public SetManifest getSetManifest(String name) throws IOException {
        return getSetManifest(getSetFile(name));

    }
    public SetManifest getSetManifest(File file) throws IOException {
        File outputDir = getOutputDir(); // context being the Activity pointer
        Log.d(getClass().getCanonicalName(), "getSetManifest: "+outputDir.getAbsolutePath());
        new ZipFile(file).extractFile("config.json", outputDir.getAbsolutePath());
        File configFile = new File(outputDir, "config.json");
        return parseManifest(file, configFile);
    }

    @Nullable
    private SetManifest parseManifest(File file, File configFile) {
        String raw = readStringFile(configFile);
        try {
            return new SetManifest(file, new JSONObject(raw));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    @NonNull
    private File getOutputDir() {
        return new File(context.getCacheDir(), UUID.randomUUID().toString() + "/");
    }

    private String readStringFile(File file){

//Read text from file
        StringBuilder text = new StringBuilder();

        try {
            BufferedReader br = new BufferedReader(new FileReader(file));
            String line;

            while ((line = br.readLine()) != null) {
                text.append(line);
                text.append('\n');
            }
            br.close();
        }
        catch (IOException e) {
            //You'll need to add proper error handling here
        }
        return  text.toString();
     }

     public Set getSet(String name) throws ZipException {
        return getSet(getSetFile(name));
     }

    private Set getSet(File setFile) throws ZipException {
        File outputDir =getOutputDir(); // context being the Activity pointer
        new ZipFile(setFile).extractAll( outputDir.getAbsolutePath());

        SetManifest manifest = parseManifest(setFile, new File(outputDir, "config.json"));
        return new Set(manifest, outputDir);
    }

    public File getSetFile(String name) {
        return new File(getSetsDirectory(), name);
    }

    public Set createSet() {

        File folder = getOutputDir();
        folder.mkdir();
        File file = new File(folder, "config.json");
        SetManifest manifest = new SetManifest(file);
        return new Set(manifest, folder);
    }

    public void save(Set set, String setName, Callback callback) {
        set.writeConfig();

        try {
            Utils.zipFolder(set.getFolder().getAbsolutePath(), new File( getSetsDirectory().getAbsoluteFile(), setName).getAbsolutePath());
            callback.onDone(null);
        } catch (IOException e) {
            e.printStackTrace();
            callback.onFail(e);
        }


    }

    public void delete(SetManifest manifest) {
        manifest.file.delete();
    }

    public void rename(SetManifest manifest, String result) {
        manifest.file.renameTo(new File(manifest.file.getParent(), result+".linka"));
    }
}

File: shared/src/androidMain/kotlin/su/linka/pictures/SetsManager.kt
package su.linka.pictures

import android.content.Context
import android.content.res.AssetManager
import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.exception.ZipException
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.File
import java.io.FileOutputStream
import java.io.FileReader
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.ArrayList
import java.util.UUID

class SetsManager(private val context: Context) {

    private val cookie = Cookie(context)

    @Throws(IOException::class)
    fun loadDefaultSets() {
        if (cookie.get(Cookie.ASSETS_LOADER, false)) {
            return
        }
        val sets = getSetsDirectory()
        val assetManager = getAssetsManager()
        val list = assetManager.list("sets") ?: emptyArray()
        for (name in list) {
            assetManager.open("sets/$name").use { input ->
                val outFile = File(sets, name)
                FileOutputStream(outFile).use { output ->
                    copyStream(input, output)
                }
            }
        }
        cookie.set(Cookie.ASSETS_LOADER, true)
    }

    fun getSetsDirectory(): File {
        val sets = File(getRootDirectory(), "sets")
        if (!sets.exists()) {
            sets.mkdirs()
        }
        return sets
    }

    private fun getRootDirectory(): File = context.filesDir

    private fun getAssetsManager(): AssetManager = context.assets

    fun getSets(): Array<SetManifest> {
        val files = getSetsDirectory().listFiles() ?: return emptyArray()
        files.sortWith { first, second ->
            when {
                first.lastModified() > second.lastModified() -> -1
                first.lastModified() < second.lastModified() -> 1
                else -> 0
            }
        }
        val manifests = ArrayList<SetManifest>(files.size)
        for (file in files) {
            try {
                if (file.isFile) {
                    manifests.add(getSetManifest(file))
                }
            } catch (error: IOException) {
                error.printStackTrace()
            }
        }
        return manifests.toTypedArray()
    }

    @Throws(IOException::class)
    fun getSetManifest(name: String): SetManifest = getSetManifest(getSetFile(name))

    @Throws(IOException::class)
    fun getSetManifest(file: File): SetManifest {
        return try {
            val configFile = extractManifestConfig(file)
            parseManifest(file, configFile)
        } catch (error: JSONException) {
            throw IOException(error)
        } catch (error: ZipException) {
            throw IOException(error)
        }
    }

    @Throws(JSONException::class, IOException::class)
    private fun parseManifest(archiveFile: File?, configFile: File): SetManifest {
        val raw = readStringFile(configFile)
        return SetManifest(configFile, JSONObject(raw), archiveFile)
    }

    private fun createWorkspace(): File {
        val root = getWorkspaceRoot()
        val folder = File(root, UUID.randomUUID().toString())
        if (folder.exists()) {
            folder.deleteRecursively()
        }
        folder.mkdirs()
        return folder
    }

    @Throws(IOException::class)
    private fun readStringFile(file: File): String {
        val text = StringBuilder()
        BufferedReader(FileReader(file)).use { reader ->
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                text.append(line)
                text.append('\n')
            }
        }
        return text.toString()
    }

    @Throws(ZipException::class)
    fun getSet(name: String): Set = getSet(getSetFile(name))

    @Throws(ZipException::class)
    private fun getSet(setFile: File): Set {
        val outputDir = createWorkspace()
        ZipFile(setFile).extractAll(outputDir.absolutePath)
        val manifest = try {
            parseManifest(setFile, File(outputDir, CONFIG_FILE_NAME))
        } catch (error: JSONException) {
            throw ZipException(error)
        } catch (error: IOException) {
            throw ZipException(error)
        }
        return Set(manifest, outputDir)
    }

    fun getSetFile(name: String): File = File(getSetsDirectory(), name)

    fun createSet(): Set {
        val folder = createWorkspace()
        folder.mkdirs()
        val file = File(folder, CONFIG_FILE_NAME)
        val manifest = SetManifest(file)
        return Set(manifest, folder)
    }

    fun save(set: Set, setName: String, callback: Callback<Any?>) {
        set.writeConfig()
        try {
            val destination = File(getSetsDirectory(), setName)
            Utils.zipFolder(set.getFolder().absolutePath, destination.absolutePath)
            set.getManifest().archiveFile = destination
            clearPreview(set.getManifest().cacheKey)
            callback.onDone(null)
        } catch (error: IOException) {
            error.printStackTrace()
            callback.onFail(error)
        }
    }

    fun delete(manifest: SetManifest) {
        manifest.archiveFile?.let {
            if (it.exists()) {
                it.delete()
            }
        }
        clearPreview(manifest.cacheKey)
    }

    fun rename(manifest: SetManifest, result: String) {
        val archive = manifest.archiveFile ?: return
        val cacheKey = manifest.cacheKey
        val parent = archive.parentFile ?: return
        val target = File(parent, "$result.linka")
        if (archive.renameTo(target)) {
            clearPreview(cacheKey)
            manifest.archiveFile = target
        }
    }

    @Throws(ZipException::class)
    fun getSetImage(manifest: SetManifest, path: String?): File? {
        if (path == null) return null
        val archive = manifest.archiveFile ?: return null
        val previewDir = ensurePreviewDir(archive, manifest.cacheKey)
        val image = File(previewDir, path)
        if (!image.exists()) {
            image.parentFile?.mkdirs()
            ZipFile(archive).extractFile(path, previewDir.absolutePath)
        }
        return image
    }

    @Throws(ZipException::class)
    private fun extractManifestConfig(file: File): File {
        val previewDir = ensurePreviewDir(file)
        val config = File(previewDir, CONFIG_FILE_NAME)
        if (config.exists()) {
            config.delete()
        }
        ZipFile(file).extractFile(CONFIG_FILE_NAME, previewDir.absolutePath)
        return config
    }

    private fun ensurePreviewDir(archive: File, cacheKey: String = archive.nameWithoutExtension): File {
        val dir = File(getPreviewRoot(), cacheKey)
        val marker = File(dir, PREVIEW_MARKER)
        val stamp = archive.lastModified().toString()
        val currentStamp = if (marker.exists()) runCatching { marker.readText() }.getOrNull() else null
        if (currentStamp != stamp) {
            if (dir.exists()) {
                dir.deleteRecursively()
            }
            dir.mkdirs()
            marker.writeText(stamp)
        } else if (!dir.exists()) {
            dir.mkdirs()
            marker.writeText(stamp)
        }
        return dir
    }

    private fun clearPreview(cacheKey: String) {
        val dir = File(getPreviewRoot(), cacheKey)
        if (dir.exists()) {
            dir.deleteRecursively()
        }
    }

    private fun getPreviewRoot(): File {
        val root = File(context.cacheDir, "previews")
        if (!root.exists()) {
            root.mkdirs()
        }
        return root
    }

    private fun getWorkspaceRoot(): File {
        val root = File(context.cacheDir, "workspace")
        if (!root.exists()) {
            root.mkdirs()
        }
        return root
    }

    private fun copyStream(input: InputStream, output: OutputStream) {
        val buffer = ByteArray(1024)
        while (true) {
            val read = input.read(buffer)
            if (read == -1) break
            output.write(buffer, 0, read)
        }
    }

    companion object {
        private const val TAG = "SetsManager"
        private const val CONFIG_FILE_NAME = "config.json"
        private const val PREVIEW_MARKER = ".source"
    }
}
